### PR TITLE
Adding a What's New component to Calypso

### PR DESCRIPTION
**Pull request by donlair**
This PR adds a new component that highlights new features. The features list is managed by a content tool available in MC (the Mobile Announcements tool).

Data for the features comes from a new endpoint added in D56831-code, and made available to Calypso via a Redux.

For the sake of testing, I added the WhatsNew component to Marketing/Tools. Designs are TBD. More context available in pbBQWj-F0-P2

#### Testing instructions
* Patch your sandbox with D56831
* Checkout this PR
* Point public-api to your sandbox
* Load the Marketing page in Calypso.

### CODE REVIEW

**Comment by simison** [comment]
Was there similar work in Gutenberg editor view as well? Just thinking if we could share the same componentry/data fetching here than there?

**Comment by donlair** [comment]
> Was there similar work in Gutenberg editor view as well? Just thinking if we could share the same componentry/data fetching here than there?

This is the [PR for What's New](https://github.com/Automattic/wp-calypso/pull/48722) in the editor.

@noahtallen is there any reason the data fetching introduced in this PR wouldn't work in the toolkit?

On the design front, I think same Guide component could work in both. @sfougnier has some concepts in a Figma, and I believe at least one of them used it.

**Comment by noahtallen** [comment]
I think this _might_ not work in ETK because ETK doesn't have access to all calypso data stores.

But I think there are some best practices for data stores which integrate across different environment, so I do think it is possible to share a data source, and so we should do that. However, I'm not quite sure what that best practice is! 😅 

@Automattic/luna could you perhaps leave some info about how the shared data stores work for gutenboarding + editor? Would the same approach work for calypso + editor?

**Comment by simison** [comment]
Cc @Automattic/team-calypso also regarding data stores in above question.

**Comment by sarayourfriend** [comment]
Would it be possible to re-write this new store using `@wordpress/data` and putting it in `@automattic/data-stores`? It would be the first time we try to use wp.data in Calypso but it might work and I _think_ putting it in the `data-stores` package would make it available to anyone wanting to consume it as long as that package can be bundled in.

**Comment by donlair** [comment]
> Would it be possible to re-write this new store using `@wordpress/data` and putting it in `@automattic/data-stores`? It would be the first time we try to use wp.data in Calypso but it might work and I _think_ putting it in the `data-stores` package would make it available to anyone wanting to consume it as long as that package can be bundled in.

I'll give it a go tomorrow to see if I can get it working!

**Comment by sarayourfriend** [comment]
@donlair Feel free to ping us if you run into any blockers 😊

**Review by tyxla — COMMENTED** [review]
The "what's new" data seems like a single API request that we never make again. Do we really want to have all the Redux boilerplate and / or create it in `data-stores`? I think having it as a simple `data-getter` should do the job just fine 😉

**Comment by simison** [comment]
@tyxla related https://github.com/Automattic/wp-calypso/pull/50016

**Comment by donlair** [comment]
@tyxla can you point me towards an example of where we pull data from a WPcom using a simple data-getter? That was my original plan, but the Calypso docs pointed me towards Redux as the preferred way of fetching and passing data to components. I'm new to all this so I wouldn't be the least bit surprised if I misinterpreted the docs.

One other factor that I _think_ justifies Redux is that in some of the design concepts for this there are ideas like tracking which announcements the user has seen and having a global button with one of those red circle indicators in the UI somewhere. Designs are still TBD though, so I haven't built the actions for those flows.

What do you think? Either way, I'd love to look into the simpler approach you mentioned so I have it in my back pocket in the future.

**Comment by jsnajdr** [comment]
> I think having it as a simple `data-getter` should do the job just fine

Even a `data-getter` caches the response so that if there are multiple consumers, they avoid calling the "what's new" REST request again. We might not need that in this case.

The simplest solution is to call
```js
wpcom.req( {
  path: `/whats-new/list`,
  apiNamespace: 'wpcom/v2'
} ).then ( whatsNewList => {
  this.setState( { whatsNewList } );
} ).catch( error => {
  this.showErrorNotice();
} );
```
in a component lifecycle or `useEffect` callback. Only when this solution is not sufficient for some reason, something higher-level and more complex can be useful.

**Comment by donlair** [comment]
thanks @jsnajdr that's helpful. Bigger picture, should my takeaway be to go with the simplest solution vs jumping through hoops to run everything through Redux?

**Comment by noahtallen** [comment]
> should my takeaway be to go with the simplest solution vs jumping through hoops to run everything through Redux?

This is a fair takeaway. In particular, since we just need to display a list of data without interacting with it, redux doesn't give us a whole lot extra. Another reason we might use redux is to share this data among multiple consumers. We probably also don't need it for that. For example, if we want to reuse this logic in a couple places (calypso + ETK), we could just share the what's new component entirely.

**Comment by razvanpapadopol** [comment]
> we could just share the what's new component entirely.

Yes, we can totally do that if the component lives in `/packages`. 
Here is an interesting discussion on this subject: p4TIVU-9vf-p2

File: client/components/whats-new/index.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * Internal Dependencies
+ */
+import QueryWhatsNewList from 'calypso/components/data/query-whats-new';
+import { getWhatsNewList } from 'calypso/state/whats-new/selectors';
+function WhatsNew( { list, isListReady } ) {
+	return (
+		<Card>
+			<div>
+				{ ! isListReady && <QueryWhatsNewList /> }
```
Comment (jsnajdr) [inline client/components/whats-new/index.js:18]: You can always render `<QueryWhatsNewList />` here without any conditional. The component itself will make sure the list is fetched only on mount.

File: client/components/whats-new/index.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import { Card } from '@automattic/components';
+/**
+ * Internal Dependencies
+ */
+import QueryWhatsNewList from 'calypso/components/data/query-whats-new';
+import { getWhatsNewList } from 'calypso/state/whats-new/selectors';
+function WhatsNew( { list, isListReady } ) {
+	return (
+		<Card>
+			<div>
```
Comment (jsnajdr) [inline client/components/whats-new/index.js:17]: Do we need this nested `div`? The `<Card />` component already renders a `div` -- why a second nested one?

**Comment by jsnajdr** [comment]
> we could just share the what's new component entirely.

Will the What's New component be reused at multiple places in Calypso or even outside the repo? There are 5000+ React components in Calypso and there should be specific reasons why any of them should be promoted to a package.

**Comment by tyxla** [comment]
> Even a `data-getter` caches the response so that if there are multiple consumers, they avoid calling the "what's new" REST request again. We might not need that in this case.
> 
> The simplest solution is to call
> 
> ```js
> wpcom.req( {
>   path: `/whats-new/list`,
>   apiNamespace: 'wpcom/v2'
> } ).then ( whatsNewList => {
>   this.setState( { whatsNewList } );
> } ).catch( error => {
>   this.showErrorNotice();
> } );
> ```
> 
> in a component lifecycle or `useEffect` callback. Only when this solution is not sufficient for some reason, something higher-level and more complex can be useful.

I agree that will also do the job well if we don't need caching, request deduplication, freshness, or any other more complicated data bits 👍  

> @tyxla can you point me towards an example of where we pull data from a WPcom using a simple data-getter? 

In case you'd like to go with a data getter because you could use the caching or request deduplication, there are plenty of examples of how we use data getters [here](https://github.com/Automattic/wp-calypso/blob/remove/notices-middleware/client/state/data-getters/index.js#L11).

**Comment by jsnajdr** [comment]
> Bigger picture, should my takeaway be to go with the simplest solution vs jumping through hoops to run everything through Redux?

Yes, I think the takeaway is that one should not need to add 7 new source files just to fetch something from a REST endpoint. That's justifiable only for complex things that are used all over the place, like REST data for user, user settings or sites.

The simplest thing is to just call `wpcom.req()` and set local state on response.

The downsides are:
- every time you navigate away from the page and back, the "what's new list" needs to be loaded again from REST. The original response was stored in local state and was lost when the owning component unmounted.
- if you want to use the "what's new" data in two components at once, there need to be two copies, as the two components don't share anything.

The `data-getter` and the "data layers" frameworks solve the caching problem, and the data getters do it in a more simple way.

If you are worried about caching and reusing the "what's new" response, the smallest upgrade from mere `wpcom.req` would be a data getter like the `requestGeoLocation` one:

https://github.com/Automattic/wp-calypso/blob/trunk/client/state/data-getters/request-geo-location.js

**Comment by noahtallen** [comment]
> Will the What's New component be reused at multiple places in Calypso or even outside the repo

I think the goal would be to use it in Calypso directly, but also from Editing Toolkit so that it's available in the editor (even away from the iframe).

**Comment by donlair** [comment]
Thanks for the feedback everybody, it's all extremely helpful! I've attempted the package + wpcom.req version in #50224.

**Comment by donlair** [comment]
This went live with the approach in #51104 instead.

**Pull request closed without merging**